src/algebra/jordan/basic.lean

```diff
@@ -0,0 +1,245 @@
+two) leaving quadratic algebras to those better versed in that theory.
+
+The conventional way to linearise the Jordan axiom is to equate coefficients (more formally, assume
+that the axiom holds in all field extensions). For simplicity we use brute force algebraic expansion
+and substitution instead.
+
+## References
+
+* [Cabrera García and Rodríguez Palacios, Non-associative normed algebras. Volume 1]
+  [cabreragarciarodriguezpalacios2014]
+* [Hanche-Olsen and Størmer, Jordan Operator Algebras][hancheolsenstormer1984]
+* [Jordan, von Neumann and Wigner, 1934][jordanvonneumannwigner1934]
+* [McCrimmon, A taste of Jordan algebras][mccrimmon2004]
+
+-/
+
+/--
+A (non-commutative) Jordan multiplication.
+-/
+class is_jordan (A : Type*) [has_mul A] :=
+(lmul_comm_rmul : ∀ a b : A, (a * b) * a = a * (b * a))
+(lmul_lmul_comm_lmul: ∀ a b : A, (a * a) * (a * b) = a * ((a * a) * b))
+(lmul_lmul_comm_rmul: ∀ a b : A, (a * a) * (b * a) = ((a * a) * b) * a)
+(lmul_comm_rmul_rmul: ∀ a b : A, (a * b) * (a * a) = a * (b * (a * a)))
+(rmul_comm_rmul_rmul: ∀ a b : A, (b * a) * (a * a) = (b * (a * a)) * a)
+
+/--
+A commutative Jordan multipication
+-/
+class is_comm_jordan (A : Type*) [has_mul A]:=
+(mul_comm: ∀ a b : A, a * b = b * a)
+(jordan: ∀ a b : A, (a * b) * (a * a) = a * (b * (a *a)))
+
+-- A (commutative) Jordan multiplication is also a Jordan multipication
+@[priority 100] -- see Note [lower instance priority]
+instance jordan_of_comm_jordan (A : Type*) [has_mul A] [is_comm_jordan A] : is_jordan A :=
+{ lmul_comm_rmul := λ a b, by rw [is_comm_jordan.mul_comm, is_comm_jordan.mul_comm a b],
+  lmul_lmul_comm_lmul := λ a b, by rw [is_comm_jordan.mul_comm (a * a) (a * b),
+  is_comm_jordan.jordan, is_comm_jordan.mul_comm b (a * a)],
+  lmul_comm_rmul_rmul := λ a b, by rw [is_comm_jordan.mul_comm, ←is_comm_jordan.jordan,
+    is_comm_jordan.mul_comm],
+  lmul_lmul_comm_rmul :=  λ a b, by rw [is_comm_jordan.mul_comm (a * a) (b * a),
+    is_comm_jordan.mul_comm b a, is_comm_jordan.jordan, is_comm_jordan.mul_comm,
+    is_comm_jordan.mul_comm b (a * a)],
+  rmul_comm_rmul_rmul := λ a b, by rw [is_comm_jordan.mul_comm b a, is_comm_jordan.jordan,
+    is_comm_jordan.mul_comm], }
+
+universe u
+
+/- A (unital, associative) ring satisfies the (non-commutative) Jordan axioms-/
+@[priority 100] -- see Note [lower instance priority]
+instance ring.to_jordan_ring (B : Type u) [ring B] : is_jordan B :=
+{ lmul_comm_rmul := by { intros, rw mul_assoc },
+  lmul_lmul_comm_lmul := by { intros, rw [mul_assoc, mul_assoc] },
+  lmul_comm_rmul_rmul := by { intros, rw [mul_assoc] },
+  lmul_lmul_comm_rmul := by { intros, rw [←mul_assoc] },
+  rmul_comm_rmul_rmul := by { intros, rw [← mul_assoc, ← mul_assoc] } }
+
+variables {A : Type*} [non_unital_non_assoc_ring A]
+
+/--
+Left multiplication operator
+-/
+@[simps] def function.End.L : A →+ add_monoid.End A := add_monoid_hom.mul
```

Comment:
I was thinking more along the lines of
```suggestion
@[simps] def function.End.L (l : A) : function.End A := (*) l
```
Which doesn't have any typeclass assumptions

Reply:
How would I define `R`? There is no `function.flip`.

Reply:
`def function.End.R (r : A) : function.End A := flip (*) r` I guess.

Reply:
So, I now need to either show or assume `has_bracket (function.End A) (function.End A)`?

Reply:
Does it not have a bracket already? It should be a ring under pointwise multiplication

Reply:
Maybe this suggestion is more trouble than it's worth

Reply:
Did you mean that I should show `instance [add_comm_group M] : ring (function.End M)`? I don't think that is mathematically true.

I can show:
```
universes uM
variables {M : Type uM}

instance [has_add M] : has_add (function.End M) :=
{ add := λ f g, λ x, f(x)+g(x), }

instance [has_zero M] : has_zero (function.End M) :=
{ zero := λ x, 0, }

lemma function_add [has_add M] (f g : function.End M) (x : M) : (f+g)(x) = f(x) + g(x) := rfl

lemma function_zero [has_zero M] (x : M) : (0 : function.End M) x = 0 := rfl

lemma function_comp (f g : function.End M) (x : M) : (f * g) x = f(g(x)) := rfl

instance [add_comm_monoid M] : add_comm_monoid (function.End M) :=
{ add := (+),
  add_assoc := λ a b c, begin
    ext,
    rw [function_add, function_add, function_add, function_add, add_assoc],
  end,
  zero := (0),
  zero_add := λ a, begin
    ext,
    rw [function_add, function_zero, zero_add],
  end,
  add_zero := λ a, begin
    ext,
    rw [function_add, function_zero, add_zero],    
  end,
  add_comm := λ a b, begin
    ext,
    rw [function_add, function_add, add_comm],
  end, }
```

However, I think `instance [add_comm_monoid M] : semiring (function.End M)` is not true because of `mul_zero` i.e. `(a*0)(x) = a(0(x)) = a(0)` but we cannot conclude that this is zero without additional conditions on `a`.

The other possibility would be to do something like this:
```
@[to_additive]
instance [has_div M] : has_div (function.End M) :=
{
  div := λ f g, λ x, f(x)/g(x),
}

@[priority 100]
instance [has_sub M] : has_bracket (function.End M) (function.End M) := ⟨λ x y, x*y - y*x⟩
```

Reply:
Thanks, I think you've convinced me that what you already have is the right approach after all. Sorry for the distraction. In that case, can you rename to
```suggestion
@[simps] def add_monoid.End.L : A →+ add_monoid.End A := add_monoid_hom.mul
```